xdesign/plot.py

```diff
@@ -367,26 +377,54 @@ def discrete_phantom(phantom, size, ratio=9, uniform=True,
     image = 0
 
     if phantom.geometry is not None and phantom.material is not None \
-       and hasattr(phantom.material, prop):
+       and np.alltrue([hasattr(phantom.material, temp) for temp in prop]):
 
-        psize = 1.0 / size
+        # select whether use a fixed pixel size (at 1) or fixed object size (1cm)
+        if fix_psize:
+            psize = 1
+        else:
+            psize = 1.0 / size
 
         # Rasterize all geometry in the phantom.
         pmin, patch = discrete_geometry(phantom.geometry, psize, ratio)
 
         # Get the property value
-        value = getattr(phantom.material, prop)(DEFAULT_ENERGY)
-
-        # Make a grid to put store all of the discrete geometries
-        image = np.zeros([size] * phantom.geometry.dim, dtype=float)
+        prop = np.atleast_1d(prop)
+        n_prop = prop.size
+        ret = [None] * n_prop
+        for i, i_prop in enumerate(prop):
+            try:
+                value = getattr(phantom.material, i_prop)(energy=energy)
+            except:
+                raise TypeError('Invalid material properties.')
+
+            # Make a grid to put store all of the discrete geometries
+            image = np.zeros([size] * phantom.geometry.dim, dtype=float)
+            imin = [0] * phantom.geometry.dim
+
+            image = combine_grid(imin, image, pmin // psize, patch * value)
+            ret[i] = image
+
+        patch = patch.astype('bool')
         imin = [0] * phantom.geometry.dim
-
-        image = combine_grid(imin, image, pmin // psize, patch * value)
+        patch = combine_grid(imin, np.zeros_like(image).astype('bool'), pmin // psize, patch)
 
     for child in phantom.children:
-        image += discrete_phantom(child, size, ratio, uniform, prop)
-
-    return image
+        temp, temp_patch = discrete_phantom(child, size, ratio, uniform, prop, energy, fix_psize, return_patch=True,
+                                            overlay_mode=overlay_mode)
+        for i in range(len(ret)):
+            if overlay_mode == 'add':
+                ret[i] += temp[i]
+            elif overlay_mode == 'replace':
+                ret[i][temp_patch] = temp[i][temp_patch]
+    if return_patch:
+        try:
+            patch += temp_patch
+        except:
+            pass
+        return np.squeeze(ret), patch
+    else:
+        return np.squeeze(ret)
```

Comment:
`ret` will be returned before it is defined if the first `if` statement is false.

This function is supposed to return "0 if phantom has no geometry or material property".